src/main/java/net/schmizz/sshj/common/KeyType.java

```diff
@@ -266,4 +322,153 @@ public static KeyType fromString(String sType) {
+
+        @SuppressWarnings("unchecked")
+        static <T extends PublicKey> Certificate<T> readPubKey(Buffer<?> buf, KeyType innerKeyType) throws GeneralSecurityException {
+            final byte[] nonce;
+            final T key;
+            final BigInteger serial;
+            long type;
+            String id;
+            List<String> validPrincipals;
+            Date validAfter;
+            Date validBefore;
+            Map<String, String> critOptions;
+            Map<String, String> extensions;
+            byte[] signatureKey;
+            byte[] signature;
+
+            try {
+                nonce = buf.readStringAsBytes();
+                key = (T) innerKeyType.readPubKeyFromBuffer(buf);
+                serial = buf.readUInt64AsBigInteger();
+                type = buf.readUInt32();
+                id = buf.readString();
+                validPrincipals = unpackList(buf.readStringAsBytes());
+                validAfter = dateFromEpoch(buf.readUInt64());
+                validBefore = dateFromEpoch(buf.readUInt64());
+                critOptions = unpackMap(buf.readStringAsBytes());
+                extensions = unpackMap(buf.readStringAsBytes());
+                buf.readString(); // reserved
+                signatureKey = buf.readStringAsBytes();
+                signature = buf.readStringAsBytes();
+            } catch (Buffer.BufferException be) {
+                throw new GeneralSecurityException(be);
+            }
+
+            Certificate.Builder<T> builder = Certificate.getBuilder();
+            builder.publicKey(key)
+                .nonce(nonce)
+                .serial(serial)
+                .type(type)
+                .id(id)
+                .validPrincipals(validPrincipals)
+                .validAfter(validAfter)
+                .validBefore(validBefore)
+                .critOptions(critOptions)
+                .extensions(extensions)
+                .signatureKey(signatureKey)
+                .signature(signature);
+            return builder.build();
+        }
+
+        static void writePubKeyContentsIntoBuffer(PublicKey publicKey, KeyType innerKeyType, Buffer<?> buf) {
+            Certificate<PublicKey> certificate = toCertificate(publicKey);
+            buf.putString(certificate.getNonce());
+            innerKeyType.writePubKeyContentsIntoBuffer(certificate.getKey(), buf);
+            buf.putUInt64(certificate.getSerial())
+                .putUInt32(certificate.getType())
+                .putString(certificate.getId())
+                .putString(packList(certificate.getValidPrincipals()))
+                .putUInt64(epochFromDate(certificate.getValidAfter()))
+                .putUInt64(epochFromDate(certificate.getValidBefore()))
+                .putString(packMap(certificate.getCritOptions()))
+                .putString(packMap(certificate.getExtensions()))
+                .putString("") // reserved
+                .putString(certificate.getSignatureKey())
```

Comment:
`putBytes` (also for other places). Esp. signatures and keys are not strings but rather `byte[]`.
Though the implementation in the buffer doesn't differ, the semantics are different.